068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,315 @@
+on how custom claims or roles are supported.
+
+The concept of roles can be directly mapped to the concept of access policies in Kafka.
+
+## Proposal
+
+Identity providers support passing custom fields in the JWT token.
+Some providers allow the data type to be string, while others allow a list of strings.
+There are not many (if any) identity providers that allow a list of objects.
+
+Therefore, in this proposal we allow the ACLs to be passed as either.
+e.g. for a list of strings:
+```json
+{
+  "acls": [
+    "cluster_x:topic_1:read",
+    ":topic2:write"
+  ]
+}
+```
+
+and for a string:
+```json
+{
+  "acls": "cluster_x:topic_1:read,topic2:write"
+}
+```
+
+The field name is configurable with default value `acls`.
+
+### Using roles to pass ACLs
+
+Sometimes the identity provider supports very fine customization of the JWT,
+but this is not always the case.
+However, pretty much all identity providers support the concept of roles or claims.
+Roles are often used to define the access level of a user or service.
+For example, the database service is given the `admin` role, while the frontend service is given the `read` role. 
+In the application layer, the role is checked to see if the user has the necessary permissions when the user tries to perform an action.
+
+We can use fine-grained roles to represent ACLs.
+Instead of naming the role `admin`, which implicitly implies all permissions, we can name the role `*:topic:*:*` or if we want to limit the permissions - `*:topic:my_topic:read`.
+This way we are using the name of the role to encode the permissions the user has,
+instead of using a generic name such as `admin` and having the application layer understand what `admin` means.
+
+As an example we can define the roles for a service A in the IDP to be:
+```json
+"roles": [
+  "*:topic:input_topic:read",
+  "*:topic:output_topic:write",
+],
+```
+Now when these "roles" are passed in the JWT, we can use them as Kafka ACLs in order to authorize service A.
+
+### ACL Syntax
+
+A single ACL has the syntax `CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS`.
+The separator `:` is chosen because it is a common separator in the Unix world.
+
+If the ACLs are passed as a string, they are separated by `,`.
+i.e., `CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS,CLUSTER_NAME:RESOURCE_TYPE:RESOURCE_SPEC:PERMITTED_ACTIONS`
+
+- `RESOURCE_TYPE` can be `topic`, `group`, `transactional_id`, `delegation_token`, `user`, or the shortened versions `t`, `g`, `tid`, `dt`, `u`.
+- `PERMITTED_ACTIONS` is a list of a subset of `read`, `write`, `create`, `delete`, `alter`, `describe`, `cluster_action`, `describe_configs`, `alter_configs`, `idempotent_write`, `create_tokens`, `describe_tokens`, `all`, or the shortened versions `r`, `w`, `c`, `d`, `a`, `de`, `ca`, `dc`, `ac`, `iw`, `ct`, `dt`. The short version for `all` is `*`.
+The list items are separated by a `+`.
```

Comment:
shortened versions are really complicated to make any sense when reading a rule imho.
Also you say using `+` as separated while in one of previous comments you were suggesting the `|`.

Reply:
I suggested to use +, as | usually is interpreted as "or", but here actually means "and".

Reply:
I decided to propose the shorthand variants only to allow users with very complex and lengthy ACL lists to be able to use this feature with as little overhead as possible. I would personally advise against using them.